File: src/corpus.test.ts
// Robustness tests over a local Java corpus (e.g. OpenJDK sources and the javac
// langtools test suite). The corpus is NOT committed (it is large and GPL); drop
// .java files under ./corpus, or point JAVA_CORPUS_DIR at a checkout:
//
//   JAVA_CORPUS_DIR=/path/to/jdk/src node --run test
//
// For every file the parser must terminate, return a SourceFile and not throw
// while binding. Diagnostics are reported (not asserted) since a heterogeneous
// corpus includes intentionally malformed compiler-test inputs. When the corpus
// is absent the suite is skipped, so CI without it still passes.

import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { dirname, join } from "node:path";
import { test } from "node:test";
import { fileURLToPath } from "node:url";

import { expect } from "expect";

import { bindSourceFile } from "./binder.ts";
import { createChecker } from "./checker.ts";
import { loadJdkStub } from "./jdkStub.ts";
import { forEachChild, parseSourceFile } from "./parser.ts";
import { createProgram } from "./program.ts";
import { type Identifier, type Node, SyntaxKind } from "./types.ts";
import { pathToUri } from "./workspace.ts";

const here = dirname(fileURLToPath(import.meta.url));
const corpusDir =
  process.env.JAVA_CORPUS_DIR ?? join(here, "..", "test-fixtures", "parser", "corpus");

function findJavaFiles(dir: string): string[] {
  const result: string[] = [];
  if (!existsSync(dir)) return result;
  for (const entry of readdirSync(dir)) {
    const full = join(dir, entry);
    if (statSync(full).isDirectory()) {
      result.push(...findJavaFiles(full));
    } else if (entry.endsWith(".java")) {
      result.push(full);
    }
  }
  return result;
}

const files = findJavaFiles(corpusDir);

test(
  "Java corpus parses without crashing",
  { skip: files.length === 0 ? "no corpus present" : false },
  () => {
    let totalDiagnostics = 0;
    let cleanFiles = 0;
    for (const file of files) {
      const source = readFileSync(file, "utf8");
      const sf = parseSourceFile(file, source);
      expect(sf.kind).toBe(SyntaxKind.SourceFile);
      bindSourceFile(sf); // must not throw
      totalDiagnostics += sf.parseDiagnostics.length;
      if (sf.parseDiagnostics.length === 0) cleanFiles++;
    }
    console.log(
      `corpus: ${files.length} files, ${cleanFiles} clean, ${totalDiagnostics} parse diagnostics total`,
    );
  },
);

// Conformance/robustness for the semantic layer: load the whole corpus into one
// Program (+ JDK stub), then resolve every identifier and type every expression.
// Real code must not crash the resolver/checker, and a healthy fraction of names
// must resolve (the rest are JDK types outside the minimal stub).
test(
  "Java corpus resolves and types without crashing",
  { skip: files.length === 0 ? "no corpus present" : false },
  () => {
    const program = createProgram();
    loadJdkStub(program);
    for (const file of files)
      program.setOpenDocument(pathToUri(file), readFileSync(file, "utf8"), 1);
    const checker = createChecker(program);

    let identifiers = 0;
    let resolved = 0;
    const walk = (node: Node): void => {
      if (node.kind === SyntaxKind.Identifier) {
        identifiers++;
        if (checker.resolveName(node as Identifier)) resolved++;
      }
      checker.getTypeOfExpression(node); // must not throw
      forEachChild(node, child => {
        walk(child);
        return undefined;
      });
    };
    for (const file of files) {
      const sf = program.getSourceFile(pathToUri(file));
      if (sf) walk(sf);
    }

    const rate = identifiers === 0 ? 1 : resolved / identifiers;
    console.log(
      `corpus semantics: ${identifiers} identifiers, ${(rate * 100).toFixed(1)}% resolved`,
    );
    expect(identifiers).toBeGreaterThan(0);
    expect(rate).toBeGreaterThan(0.5); // regression floor; real code resolves well above this
  },
);

File: togo/internal/compiler/corpus_test.go
package compiler

import (
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// Robustness tests over a local Java corpus (OpenJDK sources + the javac
// langtools test suite) under test-fixtures/parser/corpus, or JAVA_CORPUS_DIR.
// Port of src/compiler/corpus.test.ts. Every file must parse + bind without
// crashing; the semantic layer must resolve/type without crashing and resolve a
// healthy fraction of names. Skipped when the corpus is absent.

func parserCorpusDir() string {
	if d := os.Getenv("JAVA_CORPUS_DIR"); d != "" {
		return d
	}
	return filepath.Join("..", "..", "..", "test-fixtures", "parser", "corpus")
}

func parserCorpusFiles() []string {
	var out []string
	_ = filepath.WalkDir(parserCorpusDir(), func(path string, d fs.DirEntry, err error) error {
		if err == nil && !d.IsDir() && strings.HasSuffix(path, ".java") {
			out = append(out, path)
		}
		return nil
	})
	return out
}

func TestJavaCorpusParsesWithoutCrashing(t *testing.T) {
	files := parserCorpusFiles()
	if len(files) == 0 {
		t.Skip("no parser corpus present")
	}
	total, clean := 0, 0
	for _, file := range files {
		source, err := os.ReadFile(file)
		if err != nil {
			continue
		}
		sf := ParseSourceFile(file, string(source))
		if sf.Kind != SourceFile {
			t.Errorf("%s: not a SourceFile", file)
			continue
		}
		BindSourceFile(sf) // must not throw
		n := len(sf.AsSourceFile().ParseDiagnostics)
		total += n
		if n == 0 {
			clean++
		}
	}
	t.Logf("corpus: %d files, %d clean, %d parse diagnostics total", len(files), clean, total)
}

func TestJavaCorpusResolvesWithoutCrashing(t *testing.T) {
	files := parserCorpusFiles()
	if len(files) == 0 {
		t.Skip("no parser corpus present")
	}
	program := NewProgram()
	LoadJdkStub(program)
	for _, file := range files {
		if source, err := os.ReadFile(file); err == nil {
			program.SetOpenDocument(corpusURI(file), string(source), 1)
		}
	}
	checker := NewChecker(program)

	identifiers, resolved := 0, 0
	var walk func(node *Node)
	walk = func(node *Node) {
		if node.Kind == Identifier {
			identifiers++
			if checker.ResolveName(node) != nil {
				resolved++
			}
		}
		checker.GetTypeOfExpression(node) // must not throw
		node.ForEachChild(func(child *Node) bool {
			walk(child)
			return false
		})
	}
	for _, file := range files {
		if sf := program.GetSourceFile(corpusURI(file)); sf != nil {
			walk(sf)
		}
	}

	rate := 1.0
	if identifiers > 0 {
		rate = float64(resolved) / float64(identifiers)
	}
	t.Logf("corpus semantics: %d identifiers, %.1f%% resolved", identifiers, rate*100)
	if identifiers == 0 {
		t.Error("expected at least one identifier")
	}
	if rate <= 0.5 {
		t.Errorf("resolution rate %.1f%% below the 50%% regression floor", rate*100)
	}
}
